File: app/rotas/auditoria.py
"""/auditoria - trilha append-only com diff campo a campo."""

from __future__ import annotations

from fastapi import APIRouter, Request
from sqlalchemy import desc, select

from app.dependencias import SessaoDep, UsuarioDep
from app.modelos import AcessoDadoSensivel, HistoricoEvento, Usuario
from app.servicos.auditoria import (
    ConferenciaEmCurso,
    conferir_fora_da_transacao,
    janela_integra,
    ultima_conferencia,
)
from app.web import pagina

rotas = APIRouter(tags=["auditoria"])

POR_PAGINA = 100


@rotas.get("/auditoria")
def trilha(
    request: Request,
    s: SessaoDep,
    usuario: UsuarioDep,
    usuario_id: int | None = None,
    entidade: str | None = None,
    tipo_evento: str | None = None,
    pagina_num: int = 1,
    mensagem: str | None = None,
    erro: str | None = None,
):
    usuario.exigir("auditoria.ver")
    consulta = select(HistoricoEvento).order_by(desc(HistoricoEvento.id))
    if usuario_id:
        consulta = consulta.where(HistoricoEvento.usuario_id == usuario_id)
    if entidade:
        consulta = consulta.where(HistoricoEvento.entidade == entidade)
    if tipo_evento:
        consulta = consulta.where(HistoricoEvento.tipo_evento == tipo_evento)

    eventos = list(
        s.execute(
            consulta.limit(POR_PAGINA).offset((max(pagina_num, 1) - 1) * POR_PAGINA)
        ).scalars()
    )
    # A JANELA, e nao a cadeia inteira (Q-2). O passe completo percorria a tabela
    # toda a cada abertura desta tela, com o lock de escrita do SQLite na mao — a
    # ~55 us por evento, aos ~90 mil eventos uma abertura de /auditoria passava
    # dos 5 s de `busy_timeout` e derrubava a tela de todo mundo. O que a pessoa
    # tem diante dos olhos sao 100 eventos; conferir os 100 custa ~6 ms e nao
    # cresce com a idade do sistema. A cadeia inteira tem o botao abaixo.
    ok, defeito = janela_integra(s, eventos)
    tipos = sorted(
        {t for (t,) in s.execute(select(HistoricoEvento.tipo_evento).distinct()).all()}
    )
    entidades = sorted(
        {e for (e,) in s.execute(select(HistoricoEvento.entidade).distinct()).all()}
    )
    return pagina(
        request,
        "paginas/auditoria.html",
        usuario=usuario,
        eventos=eventos,
        janela_ok=ok,
        janela_defeito=defeito,
        conferencia=ultima_conferencia(s),
        tipos=tipos,
        entidades=entidades,
        filtro_usuario=usuario_id,
        filtro_entidade=entidade,
        filtro_tipo=tipo_evento,
        pagina_num=pagina_num,
        mensagem=mensagem,
        erro=erro,
        usuarios=list(s.execute(select(Usuario).order_by(Usuario.nome)).scalars()),
        acessos=list(
            s.execute(
                select(AcessoDadoSensivel).order_by(desc(AcessoDadoSensivel.id)).limit(50)
            ).scalars()
        ),
    )


# POST, e nao GET, pelo mesmo motivo do EXPORTAR-TUDO: o que sai daqui e um
# efeito colateral (uma linha nova em `conferencia_cadeia`) e uma varredura da
# trilha inteira. Com `SameSite=lax` o cookie viaja em navegacao de topo, entao
# um GET poderia ser disparado por um link de fora.
@rotas.post("/auditoria/conferir")
def conferir(request: Request, s: SessaoDep, usuario: UsuarioDep):
    """O passe completo, sob demanda. Solta a transacao antes de percorrer.

    Quem abre a trilha pode conferi-la: a pergunta "isto ainda faz prova?" e da
    mesma natureza que ler a trilha, e negar a resposta a quem ve os eventos
    seria oferecer a prova sem deixar confirma-la. O que tornava perigoso
    responde-la era o lock, e o lock saiu do caminho — `conferir_fora_da_transacao`
    comita antes de ler, e le por uma conexao que nao disputa escrita.
    """
    usuario.exigir("auditoria.ver")
    try:
        resultado = conferir_fora_da_transacao(s, usuario=usuario)
    except ConferenciaEmCurso as falha:
        return trilha(request, s, usuario, erro=str(falha))
    if resultado.integra:
        aviso = (
            f"Cadeia conferida: {resultado.eventos} evento(s), encadeamento "
            f"íntegro ({resultado.duracao_ms} ms)."
        )
        return trilha(request, s, usuario, mensagem=aviso)
    return trilha(
        request,
        s,
        usuario,
        erro=(
            f"O encadeamento não fecha a partir do evento "
            f"{resultado.primeiro_defeito_id}. Preserve o arquivo do banco como "
            "está e trate como incidente."
        ),
    )

File: nuvem/src/rotas/auditoria.ts
/**
 * /auditoria — trilha append-only com diff campo a campo.
 * Porte de `app/rotas/auditoria.py`.
 */
import { Hono } from "hono";
import { and, asc, desc, eq, type SQL } from "drizzle-orm";
import type { Ambiente, Ctx } from "../nucleo/contexto.js";
import { usuarioLogado } from "../dependencias.js";
import { acesso_dado_sensivel, historico_evento, usuario as tabela_usuario } from "../db/esquema/index.js";
import {
  ConferenciaEmCurso,
  conferir_fora_da_transacao,
  janela_integra,
  ultima_conferencia,
} from "../servicos/auditoria.js";
import type { UsuarioAtual } from "../servicos/rbac.js";
import "../servicos/pendencias.js"; // liga o sino da casca
import { pagina } from "../web.js";

export const rotas = new Hono<Ambiente>();

export const POR_PAGINA = 100;

function _inteiro(bruto: string | undefined | null): number | null {
  return bruto && /^\d+$/.test(bruto) ? Number(bruto) : null;
}

async function _trilha(c: Ctx, usuario: UsuarioAtual, recado: { mensagem?: string | null; erro?: string | null } = {}) {
  usuario.exigir("auditoria.ver");
  const tx = c.get("tx");
  // num POST (conferir) os filtros não viajam: a tela volta à primeira página
  const get = c.req.method === "GET";
  const filtro_usuario = get ? _inteiro(c.req.query("usuario_id")) : null;
  const filtro_entidade = (get && c.req.query("entidade")) || null;
  const filtro_tipo = (get && c.req.query("tipo_evento")) || null;
  const pagina_num = (get && _inteiro(c.req.query("pagina_num"))) || 1;

  const condicoes: (SQL | undefined)[] = [];
  if (filtro_usuario) condicoes.push(eq(historico_evento.usuario_id, filtro_usuario));
  if (filtro_entidade) condicoes.push(eq(historico_evento.entidade, filtro_entidade));
  if (filtro_tipo) condicoes.push(eq(historico_evento.tipo_evento, filtro_tipo));
  const eventos = await tx
    .select()
    .from(historico_evento)
    .where(and(...condicoes))
    .orderBy(desc(historico_evento.id))
    .limit(POR_PAGINA)
    .offset((Math.max(pagina_num, 1) - 1) * POR_PAGINA);
  // A JANELA, e não a cadeia inteira (Q-2): o que a pessoa tem diante dos olhos
  // são 100 eventos, e conferir os 100 não cresce com a idade do sistema.
  const [ok, defeito] = await janela_integra(tx, eventos);
  const tipos = (await tx.selectDistinct({ t: historico_evento.tipo_evento }).from(historico_evento)).map((l) => l.t).sort();
  const entidades = (await tx.selectDistinct({ e: historico_evento.entidade }).from(historico_evento)).map((l) => l.e).sort();
  return pagina(c, "paginas/auditoria.html", usuario, {
    eventos,
    janela_ok: ok,
    janela_defeito: defeito,
    conferencia: await ultima_conferencia(tx),
    tipos,
    entidades,
    filtro_usuario,
    filtro_entidade,
    filtro_tipo,
    pagina_num,
    mensagem: recado.mensagem ?? (get ? c.req.query("mensagem") : null) ?? null,
    erro: recado.erro ?? (get ? c.req.query("erro") : null) ?? null,
    usuarios: await tx.select().from(tabela_usuario).orderBy(asc(tabela_usuario.nome)),
    acessos: await tx.select().from(acesso_dado_sensivel).orderBy(desc(acesso_dado_sensivel.id)).limit(50),
  });
}

rotas.get("/auditoria", async (c) => _trilha(c, await usuarioLogado(c)));

/**
 * O passe completo, sob demanda. POST, e não GET: é um efeito colateral (uma
 * linha em `conferencia_cadeia`) e uma varredura da trilha inteira. Quem abre
 * a trilha pode conferi-la.
 */
rotas.post("/auditoria/conferir", async (c) => {
  const usuario = await usuarioLogado(c);
  usuario.exigir("auditoria.ver");
  let resultado;
  try {
    resultado = await conferir_fora_da_transacao(c.get("tx"), { usuario });
  } catch (falha) {
    if (!(falha instanceof ConferenciaEmCurso)) throw falha;
    return _trilha(c, usuario, { erro: falha.message });
  }
  if (resultado.integra) {
    return _trilha(c, usuario, {
      mensagem: `Cadeia conferida: ${resultado.eventos} evento(s), encadeamento íntegro (${resultado.duracao_ms} ms).`,
    });
  }
  return _trilha(c, usuario, {
    erro:
      `O encadeamento não fecha a partir do evento ${resultado.primeiro_defeito_id}. ` +
      "Preserve o banco como está e trate como incidente.",
  });
});
